bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java

```diff
@@ -0,0 +1,584 @@
+                        return input;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Determines the preferred voice for the currently set locale
+     *
+     * @param voices a set of voices to chose from
+     * @return the preferred voice for the current locale
+     */
+    public Voice getPreferredVoice(Set<Voice> voices) {
+        // Express preferences with a Language Priority List
+        Locale locale = localeProvider.getLocale();
+
+        // Get collection of voice locales
+        Collection<Locale> locales = new ArrayList<Locale>();
+        for (Voice currentVoice : voices) {
+            locales.add(currentVoice.getLocale());
+        }
+
+        // TODO: This can be activated for Java 8
+        // Determine preferred locale based on RFC 4647
+        // String ranges = locale.toLanguageTag();
+        // List<Locale.LanguageRange> languageRanges = Locale.LanguageRange.parse(ranges);
+        // Locale preferedLocale = Locale.lookup(languageRanges,locales);
+        Locale preferredLocale = locale;
+
+        // As a last resort choose some Locale
+        if (null == preferredLocale) {
+            preferredLocale = locales.iterator().next();
+        }
+
+        // Determine preferred voice
+        Voice preferredVoice = null;
+        for (Voice currentVoice : voices) {
+            if (preferredLocale.equals(currentVoice.getLocale())) {
+                preferredVoice = currentVoice;
+            }
+        }
+        assert (preferredVoice != null);
+
+        // Return preferred voice
+        return preferredVoice;
+    }
+
+    /**
+     * Starts listening for the keyword that starts a dialog
+     *
+     * @throws IllegalStateException if required services are not available
+     */
+    public void startDialog() {
+        startDialog(null, null, null, null, null, null, null, this.keyword);
+    }
+
+    /**
+     * Starts listening for the keyword that starts a dialog
+     *
+     * @throws IllegalStateException if required services are not available
+     */
+    public void startDialog(KSService ks, STTService stt, TTSService tts, HumanLanguageInterpreter hli,
+            AudioSource source, AudioSink sink, Locale locale, String keyword) {
+
+        // use defaults, if null
+        ks = (ks == null) ? getKS() : ks;
+        stt = (stt == null) ? getSTT() : stt;
+        tts = (tts == null) ? getTTS() : tts;
+        hli = (hli == null) ? getHLI() : hli;
+        source = (source == null) ? getSource() : source;
+        sink = (sink == null) ? getSink() : sink;
+        locale = (locale == null) ? localeProvider.getLocale() : locale;
+
+        if (ks != null && stt != null && tts != null && hli != null && source != null && sink != null) {
+            DialogProcessor processor = new DialogProcessor(getKS(), getSTT(), getTTS(), getHLI(), getSource(),
+                    getSink(), localeProvider.getLocale(), keyword);
```

Comment:
I don't know if my reading is correct:
* first we set a variable to the return variable of getVariable if it is null
* then we check if all variables differ from null
* then we use get... for every variable

Why calling the get function again?